### ISSUE TITLE
[HOLD for payment 2023-06-01] [$4000] After returning to a setting from a profile, the tooltip loses position on the profile image when you hover the mouse for the first time.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15229]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to the setting -> Profile
2. Go back to Setting and hover the mouse over the profile image.
  (hover the image quickly)

## Expected Result:
The tooltip should not lose the position even if you hover the image quickly.

## Actual Result:
 If you quickly hover over the image after returning from the profile, the tooltip loses position.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome

**Version Number:** 1.2.73-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/219512628-3d4e28ad-dabe-46d0-9be3-64bd37826d8a.mp4

https://user-images.githubusercontent.com/43996225/219512642-13a4168c-f829-4e2e-935e-cf2fce52f0ef.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676563140766109

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0175f90dca3be1912f</li>
        <li>Upwork Job ID: 1628033343922810880</li>
        <li>Last Price Increase: 2023-04-17</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @CortneyOfstad (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
@CortneyOfstad Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by CortneyOfstad** [comment]
Was OoO — tackling this now!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0175f90dca3be1912f

**Comment by MelvinBot** [comment]
Current assignee @CortneyOfstad is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor Plus for review of internal employee PR - @parasharrajat (`Internal`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MariaHCD (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Ramkrishna1998** [comment]
The tooltip position is initialized only once during the first hover and the issue occurs because of the single initialization of the position of the tooltip so even if the target element position gets changed but tooltip position will not changed. 

I have tried to implement the same scenario using the same component and I am attaching the before and after the fix 

before: https://www.loom.com/share/f40dfbbdf9f94c07bee3e3c73285de34

after: https://www.loom.com/share/fa19295843f74158a8c4164546455202

**Comment by Litande** [comment]
## Proposal
### Pleease re-state the problem that we are trying to solve in this issue.
The position of the tooltip apear in the wrong location.

### What is the root cause of that problem?
The positions is taken upon hover even if the element is in transition those it initialized there.

### What changes do you think we should make in order to solve the problem?
We can change the logic to such that the initialization of the tooltip is throttled while the cordinates are changing, with a small delay to wait for fixed cordinates.

### What alternative solutions did you explore? (Optional)
Where trying check why overlay isn't used in web display for stack navigation component in android and ios there is such overlay (transparent) which blocks interactions, but i don't think filling this for web would be an easy task.

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by MelvinBot** [comment]
📣 @Litande! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Litande** [comment]
Contributor details
Your Expensify account email: mihaelchas@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/michaelchasovin

**Comment by MelvinBot** [comment]
⚠️ Missing/invalid email or upwork profile link. Please make sure you add both your Expensify email and Upwork profile link in the format specified.

**Comment by Litande** [comment]
Contributor details
Your Expensify account email: mihaelchas@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/michaelchasovin

**Comment by MelvinBot** [comment]
⚠️ Missing/invalid email or upwork profile link. Please make sure you add both your Expensify email and Upwork profile link in the format specified.

**Comment by Litande** [comment]
Can someone help me out with Melvin here?) 
both are valid 
![image](https://user-images.githubusercontent.com/13526647/220543671-095318bc-9df7-4884-86ff-c0f15eafd482.png)
![image](https://user-images.githubusercontent.com/13526647/220543842-dcd13284-5fd3-4342-a3ab-ec6024137fb9.png)

**Comment by MelvinBot** [comment]
📣 @Prince-Mendiratta! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Litande** [comment]
@Prince-Mendiratta thanks, didn't notice )

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by ayazhussain79** [comment]
I have reported this issue, should i share my email and upwork as well.

**Comment by MelvinBot** [comment]
📣 @ayazhussain79! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by ayazhussain79** [comment]
Contributor details
Expensify account email: hussainayaz79@gmail.com

 Upwork Profile Link: https://www.upwork.com/fl/~018a62c1e70a3fab40

**Comment by Ramkrishna1998** [comment]
**Proposal**
Please re-state the problem that we are trying to solve in this issue.
The position of the tooltip appears in the wrong location.

**What is the root cause of that problem?**
The tooltip position is initialized only once during the first hover and the issue occurs because of the single initialization of the position of the tooltip so even if the target element position gets changed but tooltip position will not changed.

**What changes do you think we should make in order to solve the problem?**
We have to re-render that part after UI will get into its original position


Contributor details
Your Expensify account email: bhattramkrishna98@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0121390686e5dc8b39

**Comment by MelvinBot** [comment]
📣 @parasharrajat! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
📣 @Ramkrishna1998! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by cubuspl42** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The tooltip is first shown when the avatar is animating (although the animation is short and it's a temporary position), but stays in the incorrect position.

### What is the root cause of that problem?

The tooltip position is sampled and stored into a `fixed` CSS position.

![image](https://user-images.githubusercontent.com/2590174/220628307-29b80675-5990-4fbf-8b83-189c8e61b377.png)

### What changes do you think we should make in order to solve the problem?

There are two solution candidates that seem viable to me:

- Embed the tooltip in the DOM tree close to the avatar, with appropriate Z-index
- Reactively update the fixed tooltip position during the animation

I would first try the first approach and fall back to the second one, if the first one is difficult to achieve for some reason specific to this project and chosen dependencies.

**Comment by cubuspl42** [comment]
Contributor details
Your Expensify account email: cubuspl42.1@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~012b13cc109598f579

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by Litande** [comment]
@parasharrajat should i udate the proposal inline? or add new one ?

**Comment by Litande** [comment]
## Updated Proposal https://github.com/Expensify/App/issues/15229#issuecomment-1439511687
### Pleease re-state the problem that we are trying to solve in this issue.
The position of the tooltip apear in the wrong location.

### What is the root cause of that problem?
`ModalStackNavigator` is used in this instance and while in transition interactions on web aren't disabled,
those hovering on it, while in transition causes a trigger to show `Tooltip` which uses `Hoverable` onMouseEnter events with X and Y params that aren't in their final position.

### What changes do you think we should make in order to solve the problem?
We can change the logic to such that the initialization of the tooltip is throttled while the cordinates are changing, with a small delay to wait for fixed cordinates.

I am proposing to update `Hoverable`, in there `onMouseMove` to add an event which will also apply `IsHoverable(true)`
here
https://github.com/Expensify/App/blob/9d04b46456253955c96981db9ef38245953baf18/src/components/Hoverable/index.js#L46-L55
we should update the logic contain a throttle mechanic for acutally settings is hoverable to true so it will happen upon stationary set of mouse after, for example for 50 miliseconds to show the tooltip, this 

here is a video of the result
https://www.loom.com/share/b4f1fb67e7204659a9f280b044100611
in here the throttle is 300 but i tested with 50 and looks also good.

### What alternative solutions did you explore? (Optional)
Where trying check why overlay isn't used in web display for stack navigation component in android and ios there is such overlay (transparent) which blocks interactions, but i don't think filling this for web would be an easy task.

**Comment by Litande** [comment]
@parasharrajat updated https://github.com/Expensify/App/issues/15229#issuecomment-1440067841

Also do by some change know if there is anyone to help with joining slack send the email 3 already )

Thanks.

**Comment by cubuspl42** [comment]
@parasharrajat Would you mind commenting on my proposal? Don't you think that changing the way tooltips are layed out could solve the problem at its core?

**Comment by Litande** [comment]
@parasharrajat ofcourse,

inside `setIsHoverable`
we can add a mechanic similar to the folowing
```
if (this.timeout) {
    clearTimeout(this.timeout);
    this.timeout = null;
}
this.timeout = setTimeout(() => {
    //// SET TIMEOUT
}, 50)
```
this in fact an effect delay of hoverable exection while the mouse is moving, this is ofcourse the general idean and not how a potential fix will look like.

Also regarding the location, I am not proposing fixing the navigation this what i wrote in the optional, where we are looking into implementing the logic of cover the element with other transparent element (which blocks integrations), and also this will be only partial fix to this location if any.

Hoverable is the target of the fix/imporvement, in such a way that while the user is moving his mouse the execution of hoverable will be delayed/throttled, and will actually take effect if the user stopped on the target element and those hoverable should apply.

As for implementation in web delay effect over tooltip is a common usage take for example the "member" label on your comment hear in github, it has also a minor delay, scroll over it nothing apear scroll into it and stop and a tooltip apears.






`

**Comment by Litande** [comment]
@parasharrajat hi i just checked the code there is no delay only animation on tooltips.
 
 and in case of everywhere, we can set paramater "delay" with default as undefined in which case there won't be any.

**Comment by cubuspl42** [comment]
@parasharrajat After some experimenting, I realized that approach no. 1 doesn't integrate well with scrollables. So it's not an acceptable solution. Sorry if I made it sound confusing. I'm talking about an approach where we lay out an "anchor", having no size (size of zero; being empty) in the place we want the tooltip to point to. Then we position the tooltip absolutely to that anchor (with an extra transform, like `[{translateX: '-50%'}, {translateY: '-50%'}]` and a high z-order. It's a simple and neat solution, but such a tooltip cannot "escape" scrollables. Which makes this approach not applicable here.

I don't know why the moving tooltip is a bad thing... If what is under the tooltip moves, I'd assume that the tooltip should move too. For example, currently in Expensify when you scroll the chat while you hover an avatar, the glitch we're talking about (or a similar one) also appears. Of course, it's another topic wether the tooltip is not displayed too soon.

When you said "expensive", you mean expensive performance-wise, right?

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The tooltip position is misplaced when hover the profile image while navigation is transitioning.

### What is the root cause of that problem?
The current implementation of the tooltip uses a portal to put it at the body level. When we hover a component, it will calculate the current position of the component and the position of the tooltip based on it. However, because the transition is happening, the component position is not in a final position, yet we are calculating the tooltip position.

### What changes do you think we should make in order to solve the problem?


At first, I thought we can use `InteractionManager.runAfterInteractions`, but web does not have a full support for it yet. The alternative I could think of is 

1. to create a new function inside Navigation called `runAfterNavigation` which will execute the passed callback after the transitioning is done. We will wrap the `showTooltip` code with `runAfterNavigation`.

Here is the pseudocode of `runAfterNavigation`

```
runAfterNavigation (action):
    if isNavigating then:
        pendingAction push action
        return
    invoke action
```

If the transition is happening, we push the action to the `pendingAction` array, else we execute it immediately. We will execute all `pendingAction` later inside `setIsNavigating` when the value passed is false. The `isNavigatingValue` will be true when the transition starts and false when it ends. 

https://github.com/Expensify/App/blob/b6636348ff9ed53a4fd65ba0dba1b2f73f3809a9/src/libs/Navigation/Navigation.js#L75-L77

2. However, transition start event does not fire when we navigate forward, so we need to also call `setIsNavigating` inside `navigate` except when it's a drawer route, because drawer route never calls transition end event (only stack navigator calls the transition event [ref](https://reactnavigation.org/docs/navigation-events/)).

https://github.com/Expensify/App/blob/b6636348ff9ed53a4fd65ba0dba1b2f73f3809a9/src/libs/Navigation/Navigation.js#L153-L186

But with this 2nd change, this will prevent multiple navigation when we press quickly over several components that call `navigate`. Currently, we only prevent it when navigating back only.

**Comment by CortneyOfstad** [comment]
@MariaHCD Any updates with this? Just checking to see if we're close to selecting a proposal. Thanks!

**Comment by MariaHCD** [comment]
Catching up here. @parasharrajat let me know what your thoughts are so far and how I can help!

**Comment by MelvinBot** [comment]
@CortneyOfstad @MariaHCD @parasharrajat this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by CortneyOfstad** [comment]
@parasharrajat bump on Maria's comment above ^^^

**Comment by cubuspl42** [comment]
@parasharrajat Thank you for clarifying. I've updated my answer with many more details about the position synchronizing approach, including a code snippet and my observations about performance. I also uploaded a short video showing that this approach works in general case, instead of solving this single case.

I have a strong feeling that this solution is performant enough and I like its generality. But if you have any more thoughts or suggestions about performance, I could try to improve the proposal or do more testing.

**Comment by CortneyOfstad** [comment]
@MariaHCD and @parasharrajat thoughts on @cubuspl42 proposal above? ^^^

**Comment by Litande** [comment]
> @parasharrajat Hi i went over the layers of Hoverable and Tooltip and didn't find delay mechanics, also check the behavior and seems to only have small animation sequence, can you please check it out?
> 
> Regrading evertwgere, i am proposing for it be parametrized so it can be adjusted depending on the implementation, we can set paramater "delay" with default as undefined in which case there won't be any.
> 
> i have also went over where it is implemented, and it exists in 5 components mainly tooltips, and report user namings.

@parasharrajat 

can please check regarding the delay i didn't find it anywhere...

**Comment by Litande** [comment]
@parasharrajat you are right i missed, it.

after inspection i noticed this delay isn't actually happening constantly in most cases the first option is happening without any delay in the case here the first options, the `TooltipSense` is a global state executing debaunce logic for: as it seems imidiate apearance management of the tooltip implemented some time ago, i may be loosing something in understanding of this element.

but as it seems the responsiblility of delay shouldn't have been on the presentation component (`Tooltip`) but on the HOC `Hoverable` which is responsible for showing/hiding the element it wraps. including debouncing hiding/showing the tooltip in case the user removed his mouse of the item and returned it back.

regarding `TooltipSense` i can see it was your PR, if you give more insight into it's target, i can better understand the eco and give a better proposal in this case. 

Thanks.

**Comment by CortneyOfstad** [comment]
Thank you for the update @parasharrajat!

**Comment by CortneyOfstad** [comment]
@parasharrajat when you have a chance, can you provide an update? Thanks!

**Comment by MelvinBot** [comment]
@CortneyOfstad @MariaHCD @parasharrajat this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Internal assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $4000

**Comment by Litande** [comment]
@parasharrajat in scope of the delay i also suggest to move the current delay to hoverable and to hide the element with the same delay those (with cancel of hide if the user returns to it those fullfiling the TooltipSense in scope of each hoverable)

Thanks.

**Comment by cubuspl42** [comment]
@fedirjh Do I understand correctly, the the proposed approach for forwarding the `ref` to a `TextInput` inside `react-native-picker/picker` was intended to be done mostly (or fully) in our `react-native-picker-select` fork?

What was your perspective regarding filing a PR in the upstream? Did you intend to do that? I would like to follow your intention in this aspect.

**Comment by MelvinBot** [comment]
Current assignee @CortneyOfstad is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @MariaHCD is eligible for the External assigner, not assigning anyone new.

**Comment by Litande** [comment]
@parasharrajat Good Morning

Currently the `TooltipSense` sense is global and if we have 2 tooltips near one another or on same location like in the issue (The Avatar)

The behavior applies to both of them, so if it is already activated the other tooltip or itself will apear imidiatly without delay or animation, please correct me if you mean this the cases when we want the tooltip to apear imidiatly and not delayed.
the logic is preset here 
https://github.com/Expensify/App/blob/9d04b46456253955c96981db9ef38245953baf18/src/components/Tooltip/index.js#L108-L119
this logic impact cross all tooltips if this is by design then moving the delay to `Hoverable` element will break this logic that why i asked if that is by design, moving it will change the behavior that the delay will be localized to each hoverable and not shared globaly, i commented regrading imidiate reapearance of the tooltip using "hide" delay logic but again with implementation in `Hoverable` it will be in each element independetly. 

I would like to note: this doesn't introduce any new static numbers (i don't believe in them) it uses the one currently located in tooltip.

but it does impact other components like `Banner` `OptionRow` `ReportActionItem` `Footer` `OptionRowLHN` in total 6 components including `Tooltip` those the default behavior can be without delay and if `Hoverable` is passed a `delay` prop it implements throttling using the same delay logic, this solution through to work good will require also to pass `TooltipSense` to the `Hoverable` component in order to ignore delay in case `TooltipSense` is active, it will also require to add on back button a call to `TooltiSense.deactivateImidiate` to remove the debounce those resetting `TooltipSense` to preserve it's current logic.

I already implemented and tested it, can present code/video if you would like to see it.

**Comment by Litande** [comment]
@parasharrajat yes i understood already, that's why my last comment includes logic to preserve it.

**Comment by Litande** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After returning to a setting from a profile, the tooltip loses position on the profile image when you hover the mouse for the first time.
### What is the root cause of that problem?
There 2 events that are causing this issue.
1. The back button on settings have also a tooltip which triggers active state on `TooltipSense` component.
2. While the initial display in `Tooltip` is delayed (500 ms - atm) when `TooltipSense` is active it is happening imidiate, those the positions of tooltip is passed imidiatly and the tooltip is shown in the wrong place


### What changes do you think we should make in order to solve the problem?
I propose 2 things:
1. Extend `TooltipSense` component to have method called `deactivateImidiate` which will reset active state on back button click.
2. In `Tooltip` to move the `delay` parameter as depency on active state before the execution of `getWrapperPosition` which desides on the tooltip position
this will effectly will take the correct position and conserve the TooltipSense behavior.

### What alternative solutions did you explore? (Optional)
Moving the delay to `Hoverable` component but at this point i belive it is over complecated without breaking the `TooltipSense` behavior.

**Comment by Litande** [comment]
@parasharrajat i have added a new [proposal](https://github.com/Expensify/App/issues/15229#issuecomment-1481037962) which is to me seems to be much less complicated then the previous one.

Thanks.

**Comment by Litande** [comment]
@parasharrajat can you lease check my new [proposal](https://github.com/Expensify/App/issues/15229#issuecomment-1481037962)? Thanks

**Comment by CortneyOfstad** [comment]
@parasharrajat @MariaHCD Where are we at here? Just checking for an update — TIA!

**Comment by cubuspl42** [comment]
@parasharrajat There is another issue that proves that the tooltip that "levitates" after the view is scrolled is also considered an issue in Expensify. I believe that it's a big argument toward a systematic solution!

**Comment by cubuspl42** [comment]
## Proposal

@parasharrajat
After some more experimentation, I'm ready to file the perfect proposal 😃

### Please re-state the problem that we are trying to solve in this issue.

The tooltip doesn't stay on top of the tooltipped element, but stays in viewport-relative position where the cursor initially "touched" the target element.

### What is the root cause of that problem?

The position of the tooltipped element is sampled once, at the time the tooltips was first meant to be shown, and stored in a `fixed` CSS position.

### What changes do you think we should make in order to solve the problem?

We should keep tooltips on top of the target element.

I previously proposed a solution, that depended on the animation frames to do that. While it's true that I didn't observe any performance problems in practice, I do agree that it's not a perfect solution. Perfectly, we'd like to _observe_ the position of the target element and update the position of the tooltip only when it changes.

As tooltips are a cursor-related thing, we support them only on Web (and Desktop, which uses Web as the backend via Electron).

On the Web, we'd like to use something like `BoundingClientRectObserver`, similar to [`ResizeObserver`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) and other members of the `*Observer` family.

Unfortunately, something like that currently doesn't exist in the standard.

What's the good news is that I figured out that it can be easily "polifilled", so it handles a big number of cases, including ours! The implementation depends on [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) and [`ResizeObserver`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver).

The usage of that class that syncs the position of two elements looks like this:

```js
new BoundingClientRectObserver((bounds) => {
    targetElement.style.left = `${bounds.x}px`;
    targetElement.style.top = `${bounds.y}px`;
}).observe(source);
```

The implementation of this class that handles simple cases (like ours!) is one file of straightforward JavaScript. 

I propose to:
- create a micro-package `bounding-client-rect-observer` on NPM, that could additionally attract attention to the Expensify project
- use it to sync the tooltip's position with the tooltipped element's position
- file a WHATWG proposal for such observer

This is a preview (in the slow motion), where I synced a position of a red rectangle appended to `body` with the avatar's position. We can see, that it basically has **zero delay**, while the `measureLayout` has a slight delay (not a very big deal, but it exists).

https://user-images.githubusercontent.com/2590174/229113247-1d3991dd-1492-420f-b20b-b230015b6513.mp4

### What alternative solutions did you explore?

- Not creating a NPM package, if that's a formal/licensing problem for us

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

After returning to a setting from a profile, the tooltip loses position on the profile image when you hover the mouse for the first time.


### What is the root cause of that problem?

When the animation is not over, `ToolTip.this.wrapperView.measureInWindow` cannot measure the exact position


### What changes do you think we should make in order to solve the problem?

We'd better put the position measurement part after the animation
https://github.com/Expensify/App/blob/27dca97f7ba582b16d59944fda33f22c73065566/src/components/Tooltip/index.js#L69-L71
``` diff
+ InteractionManager. runAfterInteractions(() => {
        this.wrapperView.measureInWindow((x, y, width, height) => resolve({
             x, y, width, height,
        }));
+ })
```

https://github.com/necolas/react-native-web/blob/985c79f9247b53fce7a728d68c6cacebd506ef2c/packages/react-native-web/src/exports/InteractionManager/index.js#L61-L66
The implementation of `InteractionManager` has been merged upstream, we need to merge it into `Expensify/react-native-web/InteractionManager`


### What alternative solutions did you explore? (Optional)
Not Yet

**Comment by MariaHCD** [comment]
Reviewing the timeline on this issue, I'm reverting the price back to $4k because there have been proposals here but these just needed to be fully reviewed. 

In any case, we are going to stop automatic price increases on issues: https://expensify.slack.com/archives/C01GTK53T8Q/p1680387336537559?thread_ts=1680354475.602609&cid=C01GTK53T8Q

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $4000

**Comment by rojiphil** [comment]
@parasharrajat 
Please find below my proposal. Kindly let me know your comments on the same.

### **Proposal**

### **Please re-state the problem that we are trying to solve in this issue.**
After returning to a setting from a profile, the tooltip loses position on the profile image when you hover the mouse for the first time

### **What is the root cause of that problem?**
The tooltip loses position on the profile image as the tooltip location is determined right at the point when the mouse pointer enters the element. Here, we assume that the element for which the tooltip is displayed is already in the final position which may not be the case always. Hence, the problem.

### **What changes do you think we should make in order to solve the problem?**
1) Implement onmousemove handler and call showToolTip so that the tooltip get's positioned when the element is in the final position.

[src\components\Tooltip\index.js]

```
handleMouseMove(mouseEvent) {
         this.showTooltip();
}  
```   

```
render() {
        let child = (
            <View
                onMouseMove={this.handleMouseMove}
            >
```
2) Use _.debounce to postpone the execution of the handler (showToolTip) until 200ms has elapsed.
3) To ensure that hideToolTip does not fall out of sequence, also use _.debounce for hideToolTip until 200ms has elapsed. 
[src\components\Tooltip\index.js]

```
        this.showTooltip = _.debounce(this.showTooltip.bind(this), 200, false);
        this.hideTooltip = _.debounce(this.hideTooltip.bind(this), 200, false);
        this.handleMouseMove = this.handleMouseMove.bind(this);
```

### **What alternative solutions did you explore? (Optional)**
**

**Comment by MelvinBot** [comment]
📣 @rojiphil! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by rojiphil** [comment]
Contributor details
Your Expensify account email: roji@owesee.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0166657026273cb5b2

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by MelvinBot** [comment]
@CortneyOfstad, @MariaHCD, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by rojiphil** [comment]
> Going to check this today.

@parasharrajat Did you get a chance to check this [proposal](https://github.com/Expensify/App/issues/15229#issuecomment-1494810571)

**Comment by MelvinBot** [comment]
@CortneyOfstad, @MariaHCD, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@CortneyOfstad, @MariaHCD, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by CortneyOfstad** [comment]
Sorry, was OoO last week. @parasharrajat where are we at with this?

**Comment by MelvinBot** [comment]
@CortneyOfstad, @MariaHCD, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by CortneyOfstad** [comment]
@parasharrajat Any updates on this?

**Comment by CortneyOfstad** [comment]
Re-assigning as I am heading out on sabbatical 👍

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MariaHCD (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @MariaHCD is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @anmurali is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cead22 (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by flodnv** [comment]
Thank you for sharing your thoughts @cubuspl42. I agree with what you said. @parasharrajat told me he reviewed all of the issue and will post an update soon. Thank you for your continued patience.

>I agree that the tooltip thing might not be a very big problem and could potentially be labeled as "WON'T FIX." 

In the future, let's try to be more critical of these issues and try to not fall into the traps that this kind of super corner case bugs can lead us into, by closing them early (or even not reporting them) 😄

**Comment by rojiphil** [comment]
@parasharrajat Can you please review my proposal too? It is already mentioned [here](https://github.com/Expensify/App/issues/15229#issuecomment-1494810571)

**Comment by MelvinBot** [comment]
@flodnv, @anmurali, @parasharrajat Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by flodnv** [comment]
@parasharrajat indicated he will have this reviewed by tomorrow : https://expensify.slack.com/archives/C01GTK53T8Q/p1682954451631219?thread_ts=1682369879.025089&cid=C01GTK53T8Q

**Comment by rojiphil** [comment]
@parasharrajat The reason why the tooltip is not correctly positioned is that we are not waiting for the tooltip's wrapper view to stabilize its position on the web browser. Using the mouse move event handler on the wrapper view will ensure that showTooltip is called only after the tooltip's wrapper view is stabilized in its position with respect to the mouse pointer location. This will help us to get the desired position of the tooltip. Hope this helps in understanding how the [proposal](https://github.com/Expensify/App/issues/15229#issuecomment-1494810571) solves the issue. IMHO, this implementation will solve the problem with few lines of code and is a straightforward approach as compared to other approaches. 

Sharing the [testing branch](https://github.com/rojiphil/App/tree/15229-tooltipLocation) for you to check code changes and run tests if required. I did test the mentioned problem and it works well. Looking forward to getting your feedback too on this.

**Comment by Litande** [comment]
@parasharrajat can you please review my updated proposal, you didn't mention anything regarding it, [update proposal](https://github.com/Expensify/App/issues/15229#issuecomment-1439511687)

**Comment by Litande** [comment]
@parasharrajat hi, as you mentioned a few times, there is already a "Delay"
https://github.com/Expensify/App/blob/9d04b46456253955c96981db9ef38245953baf18/src/components/Tooltip/index.js#L115
but this delay happens for the animation as i mentioned in the explanation in point 2. and the time of apearance the tooltip have already recevied the wrong positions, i will make again a remark, i am not proposing to add anything new in this step but to move this delay where it won't cause position calculation before executing the actual animation.

Regarding "polluting" 
https://github.com/Expensify/App/blob/d4ed2335e919b1a3ded2cb2705daae0e311c50f7/src/components/Tooltip/TooltipSense.js#L13-L16
this block written by you is actually a function that sets the TooltipSense to imidiate activation and canceles deactivation debaunce, i am only proposing to add the oposite function for imidiate deactivation 

Regarding the third party button, while in general i agree the thing is this is directly related to the issue in question the back button is also wrapped by a tooltip those activating TooltipSense, if using a third party isn't correct to solve the localized issue, then a better approache would be to set on click of Tooltip to call `deactivateImidiate` those still solving the issue, without "third-party"

**Comment by Litande** [comment]
@parasharrajat it can be in qoute a few places, but i believe a better location may be 
https://github.com/Expensify/App/blob/9d04b46456253955c96981db9ef38245953baf18/src/components/Tooltip/index.js#L65-L77 it is already promise but in case `!TooltipSense.isActive()` the it should be delayed and the delay i mentioned above will be remove, this effectivly moves the delay.

regarding the click, we actually click the content of Tooltip for example of this issue, we click the back button
https://github.com/Expensify/App/blob/d4ed2335e919b1a3ded2cb2705daae0e311c50f7/src/components/HeaderWithCloseButton.js#L164-L176

We can add this under the "onBlur"
https://github.com/Expensify/App/blob/d4ed2335e919b1a3ded2cb2705daae0e311c50f7/src/components/Tooltip/index.js#L172
onClick capture to intercept child (wrapped by tooltip) items and here to deactivate the TooltipSense

This is much more generic solution then putting on the back button just to solve this issue.

**Comment by hellohublot** [comment]
@parasharrajat  Thanks, you can try my fork, if it still doesn't work, please feel free to ask me

```diff
-    "@expensify/react-native-web": "0.18.15",
+    "@expensify/react-native-web": "https://github.com/hellohublot/Expensify-react-native-web-npm.git#7a907a14d83bb8f3df5942d50aa172b5ed914e19",
```

**Comment by cubuspl42** [comment]
@parasharrajat 

> I noticed that your solution has three bugs.

Oh crap. Thanks for finding this!

From a very quick investigation, it appears that these three points have the same root cause:

```tsx
<div ref={this._root}>
    {this.props.children}
</div>
```

I'm wrapping the provided child in an extra div, which probably wasn't really a good idea, as I think about it now. The wrapper div can have a different size than its inner child, which is reflected by incorrect tootlip positioning in some cases.

The reason I went this way is because [`findDOMNode`](https://react.dev/reference/react-dom/findDOMNode) is deprecated in React.

But at the same time, I believe that `BoundsObsever` matches the single use case that [the docs consider acceptable](https://react.dev/reference/react-dom/findDOMNode#adding-a-wrapper-div-element):

> Sometimes a component needs to know the position and size of its children. This makes it tempting to find the children with findDOMNode(this), and then use DOM methods like [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) for measurements.
>
> There is currently no direct equivalent for this use case, which is why findDOMNode is deprecated but is not yet removed completely from React.

I'll replace the wrapper `div` with `findDOMNode`, and I believe that this should solve the problems with incorrect tooltip positioning.

> We can make the scroll listeners passive In bound-observer so that they slow down the scrolling.

That's a good idea! I didn't think of that. I'll apply this suggestion.

It's evening in my time zone, so I'll apply both changes to my branch first thing tomorrow morning.

**Comment by Litande** [comment]
@parasharrajat i would like to mention regarding the scrolling that i gave a [proposal](https://github.com/Expensify/App/issues/15757#issuecomment-1489147581) there the difference is in a single listener to scroll event on which `deactivateImidiate` should be also called, in which case it also resolves that issue, i tested the behavior and it similar to that we can find in slack application for example.
Thanks.

**Comment by cubuspl42** [comment]
@parasharrajat I've pushed changes to my branch. Versions are bumped, so run `npm install`.

After using `findDOMNode` instead of a wrapper `div`, tooltips that were previously misaligned are now aligned correctly.

I've also made the scroll event listener passive, which might slightly increase performance.

**Comment by flodnv** [comment]
Thanks for all the discussions so far. I agree that @cubuspl42's proposal is the best because it fixes the root cause of the issue holistically, as proven by the fact that it also fixes another bug. I think using the deprecated findDOMNode API is okay, for the reasons mentioned in the comment above this one.

**Comment by flodnv** [comment]
📣 @cubuspl42  You have been assigned to this job by @flodnv!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by flodnv** [comment]
Awesome thanks, yes it should work!

**Comment by Litande** [comment]
@parasharrajat "https://github.com/Expensify/App/issues/15229#issuecomment-1439511687 Will not work in all cases. Depends on the tooltip click or different button(back button) to trigger things that will not work all the time." in what cases won't it work i gave a simple solution for any click, and also gave a solution to the second issue.

**Comment by flodnv** [comment]
@parasharrajat and myself placed the PR on a short HOLD whilst we are reviewing the proposed solution more in depth with help of additional engineers. Internal link: https://expensify.slack.com/archives/C02NK2DQWUX/p1684183945133759

I aim to have an update in the next 24-48 hours.

**Comment by flodnv** [comment]
At @Litande's request, we took more time to review their proposal compared to @cubuspl42's proposal -- which I should note is something we normally do not do after weeks of reviewing proposals and finally accepting one. 

We ran it by the wider internal and C+ team and requested the objective opinion of a few people, including some that are heavily involved in our React Native app. 

We unanimously agree to go forward with the accepted proposal from @cubuspl42 -- please resume work on your PR.

@Litande ultimately, your proposal does not address the root cause as elegantly as @cubuspl42's proposal does, specifically due to the use of `setTimeout`. As a reminder from our contributor guidelines, personal time or resources applied towards investigating a proposal will not guarantee compensation. Compensation is only guaranteed to those who [propose a solution and get hired for that job](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job). However, we appreciate your effort and are staying open to any other proposals you may have on other issues. 

Thank you for everyone's patience and understanding resolving this issue.

**Comment by Litande** [comment]
@flodnv 

I would like to note this setTimeout isn't new it was only moved from delay property of the animation.

But that at least a descent reason for proposal rejection. 

Thanks

**Comment by anmurali** [comment]
@parasharrajat can you complete the checklist above so I can pay and add regression (if applicable) before closing?

**Comment by flodnv** [comment]
It sounds super edge case, I'm not sure TBH. What do you think @anmurali ? Should we ask QA?

**Comment by anmurali** [comment]
No harm in adding a regression test proposal in a GH for Applause to evaluate.

**Comment by anmurali** [comment]
Also sent offers to @cubuspl42 and @parasharrajat for payment

**Comment by ayazhussain79** [comment]
@anmurali I reported that bug

**Comment by anmurali** [comment]
Sent offer

**Comment by flodnv** [comment]
Are we done here @anmurali ?

**Comment by anmurali** [comment]
Paid!

**Comment by ayazhussain79** [comment]
@anmurali could you please close this bug contract on upwrok for me
